flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -243,7 +249,23 @@ void requestPartitions() throws IOException, InterruptedException {
 				}
 
 				for (InputChannel inputChannel : inputChannels.values()) {
-					inputChannel.requestSubpartition();
+					executor.submit(() -> {
+						try {
+							inputChannel.initializeState(reader);
+						} catch (Throwable t) {
+							inputChannel.setError(t);
+						}
+					});
```

Comment:
I would move the for loop inside the executor, to avoid race conditions and not depend on `executor` being single threaded.

Reply:
good point